File: 2025/03/solution.py
lines = []

lines = open("input.txt", "r").read().splitlines(False)


def solve(digits):
    result = 0

    for line in lines:
        length = len(line)
        selected = []
        start = 0

        for i in range(digits):
            end = length - (digits - 1 - i)
            best_idx = start

            for j in range(start, end):
                if line[j] > line[best_idx]:
                    best_idx = j

            selected.append(line[best_idx])
            start = best_idx + 1

        top = int("".join(selected))
        result += top

    return result


def part1():
    return solve(2)


def part2():
    return solve(12)


if __name__ == "__main__":
    print(f"Solution part 1: {part1()}")
    print(f"Solution part 2: {part2()}")

File: 2025/03/solution.go
package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func main() {
	// How many digits in the "best suite"
	DIGITS := 12

	body, err := os.ReadFile("input.txt")

	if err != nil {
		log.Fatalf("unable to read file: %v", err)
	}

	lines := strings.Split(string(body), "\n")

	result := 0

	for _, line := range lines {
		line = strings.TrimRight(line, "\r")

		if len(line) == 0 {
			continue
		}

		length := len(line)
		selected := make([]string, 0, 12)
		start := 0

		for i := range DIGITS {
			end := length - (DIGITS - 1 - i)
			best_idx := start

			for j := start; j < end; j++ {
				if line[j] > line[best_idx] {
					best_idx = j
				}
			}

			selected = append(selected, string(line[best_idx]))
			start = best_idx + 1
		}

		top, err := strconv.Atoi(strings.Join(selected, ""))

		if err != nil {
			fmt.Println(err)
		}

		result += top
	}

	fmt.Println("Solution:", result)
	os.Exit(0)
}
